lib/rspec/mocks/proxy.rb

```diff
@@ -98,11 +98,16 @@ def replay_received_message_on(expectation, &block)
 
       # @private
       def check_for_unexpected_arguments(expectation)
-        @messages_received.each do |(method_name, args, _)|
-          next unless expectation.matches_name_but_not_args(method_name, *args)
+        return if @messages_received.empty?
 
-          raise_unexpected_message_args_error(expectation, *args)
-        end
+        # Check if any of the arguments were expected.
+        return if @messages_received.any? do |(method_name, args, _)|
+                    !expectation.matches_name_but_not_args(method_name, *args)
+                  end
+
+        unexpected_args = @messages_received.map { |_, args, _| args }.uniq.flatten
```

Comment:
The `uniq.flatten` here concerns me.  With this in place, consider this example:

``` ruby
it 'groups the "got" arguments based on the method call that included them' do
  dbl = double(:expected_method => nil)
  dbl.expected_method(:one, :four)
  dbl.expected_method(:two, :four)
  dbl.expected_method(:one, :four)

  expect(dbl).to have_received(:expected_method).with(:three, :four).once
end
```

...which fails with:

```
     Failure/Error: expect(dbl).to have_received(:expected_method).with(:three, :four).once
       Double received :expected_method with unexpected arguments
         expected: (:three, :four)
              got: ([:one, :four, :two, :four])
       Diff:
       @@ -1,2 +1,2 @@
       -[:three, :four]
       +[[:one, :four, :two, :four]]
```

...which is quite confusing.  It makes it seem that one call was received with those four arguments.  Instead, I'd like to see something like:

```
Failure/Error: expect(dbl).to have_received(:expected_method).with(:three, :four).once
  Double received :expected_method with unexpected arguments
    expected: (:three, :four)
         got: (:one, :four) (2 times)
              (:two, :four) (1 time)
```

(And perhaps two separate diffs, one per call...or maybe no diff at all; I'm not sure yet.)